Unified login/logout methods

Wauplin opened the pull request:
Fix https://github.com/huggingface/huggingface_hub/pull/1098.

This PR adds the login and logout methods as the main (only) entrypoint to login/logout in `huggingface_hub`. Most of the code is a copy from the existing code that was living in `huggingface_hub.commands.user` (which is now cleaner, with only CLI-related code).

I did not change any behavior on the git credential sutff (discussed in https://github.com/huggingface/huggingface_hub/issues/1051) as it is not the focus of this PR. Also unrelated from discussions happening in https://github.com/huggingface/huggingface_hub/issues/1094.


### Example:
```py
from huggingface_hub import login, logout

login()  # displays a widget in a notebook, a prompt in terminal otherwise

login("hf_xxx") # non-blocking login

logout() # logout completely
```

### Still compatible:
```py
from huggingface_hub import notebook_login

notebook_login()  # same as before
```

```
huggingface-cli login
huggingface-cli logout
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1111). All of your documentation changes will be reflected on that endpoint.

julien-c commented on src/huggingface_hub/utils/_runtime.py:
```diff
@@ -142,3 +142,39 @@ def is_torch_available() -> bool:
+    _is_google_colab = False
+
+
+def is_notebook() -> bool:
+    """Return `True` if code is executed in a notebook (Jupyter, Colab, QTconsole).
+
+    Taken from https://stackoverflow.com/a/39662359.
+    Adapted to make it work with Google colab as well.
+    """
+    try:
+        shell_class = get_ipython().__class__  # noqa: F821
+        for parent_class in shell_class.__mro__:  # e.g. "is subclass of"
+            if parent_class.__name__ == "ZMQInteractiveShell":
+                return True  # Jupyter notebook, Google colab or qtconsole
+        return False
+    except NameError:
+        return False  # Probably standard Python interpreter
+
+
+def is_google_colab() -> bool:
+    """Return `True` if code is executed in a Google colab.
+
+    Taken from https://stackoverflow.com/a/63519730.
+    """
+    return _is_google_colab
```
nice, those are likely to be helpful at some point

Wauplin replied:
yep :)

julien-c reviewed (commented):
LGTM!

Question: are we sure we want to merge notebook_login and login(token=..) into one? 

I think it makes sense, but just want to make sure everyone it ok with this (cc @sgugger too)

LysandreJik reviewed (approved):
I would tend to err towards having separate methods, `login` and `notebook_login`. Sure, it can be handled by a single `login`, but it feels super magical to have it behave differently when passing the token vs when not.

I would favor having a single method if code could be reused across platforms; but when passing from a script to a notebook, there will need to be manual intervention to edit the code to fit the notebook anyway, so I don't think it's particularly useful to bundle in a single method.

Other than that, the PR looks good to me!

Wauplin commented:
Here were my thoughts when building a single `login` method:
- I feel it ease the user experience e.g. "you just have to know that there is 1 entrypoint to login" => it's more concise when sharing snippets of code in documentation or social media. No need to explain "if you have a token, use `login(...)` if you don't have a token and are in a notebook use `notebook_login()`, if you don't have a token and are in a script use `interpreter_login()`" (<- this one is not mandatory-)
- If someone copies a snippet found on the internet (stackoverflow, documentation, existing notebook...), it doesn't need to adapt the `from huggingface_hub import notebook_login` to `from huggingface_hub import login`. Same if you take over/copy an entire notebook from someone else and want a non-blocking login you just need to paste your token as argument without changing anything else.
- code doesn't have to be adapted when you convert your notebook to a script with [jupyter nbconvert --to script my_notebook.ipynb](https://nbconvert.readthedocs.io/en/stable/usage.html#convert-script)
- a downstream library can also benefit from it. Let's say you implement a lib that train something and push to the hub. You could have some logic like "if not logged in, then use `login()` before pushing" -> the downstream lib doesn't have to know if  the end user is in a notebook or a terminal

That's said, I fully agree that those are only marginal benefits. If you feel it's too magical, I'm fine with reverting back to have separate login methods.
(also cc @osanseviero for opinion)

julien-c commented:
(BTW just linking to [internal convo](https://huggingface.slack.com/archives/C02V5EA0A95/p1665585096039299))

julien-c commented:
is there any drawback to have this new `login()` behavior but still keeping (and documenting) a `notebook_login` too? 

(which is what this PR does anyways, IIUC)

^ Personally I'd tend to do this, but I also don't feel that strongly about it

Wauplin commented:
> (which is what this PR does anyways, IIUC)

Yes, that's the case. I think keeping `notebook_login` is good just because it is widely used and known. I removed it from documentation (in favor for `login()`) but I can put it back, at least have it in the Reference section.

julien-c commented:
to me re-adding to doc would be a good solution, yes. But yeah let's wait for others to chime in =)

osanseviero commented:
Yesterday I thought a single method for those very different developer experiences might lead to confusion and sometimes appear too much as a black box/magic.  A `notebook_login` makes it much clearer that the behavior is different, getting a widget is very different.

But after reading all these points, I think having this behavior + a still well-maintained `notebook_login` would make sense. I'm still not 100% convinced, though. I played with it in the colab and it's nice

sgugger reviewed (approved):
Thanks for working on this! As a note, `notebook_login` should never be fully removed IMO and its equivalent `interpreter_login` should be made public, as it's quite hard to identify for sure we are in a notebooks those days, with all the new options on top of Jupyter and Colab (VsCode notebooks, some weird stuff in data bricks...). From experience in Transformers, the test might get the wrong value, so leaving a way to the user to choose the function they want (while still leaving a unified `login` as you did) is helpful.

As a note: the Codecov util make reviewing PR very difficult (and much longer for the reviewer) with all the comments. You should consider disabling it.

Wauplin commented:
Thanks everyone for the feedback. @sgugger I kept your suggestion to make both `notebook_login` and `interpreter_login` official so that the user can always force the use of the widget (or not). I have updated the documentation/docstrings accordingly. I think we are now good to go :tada:

--- 

> As a note: the Codecov util make reviewing PR very difficult (and much longer for the reviewer) with all the comments. You should consider disabling it.

@sgugger understood. I agree that when code is not tested (like in this PR) it is very annoying. [I changed the config](https://github.com/huggingface/huggingface_hub/pull/1111/commits/04a8063965e9a7ca5b5865be298b4e9f88a95555) to disable the comments in the "diff section". For your info (I just learnt about it from [the doc](https://docs.codecov.com/docs/github-checks#hiding-annotations-in-the-files-view)), it is possible to toggle them by pressing "a" when you review the code in [Files changed](https://github.com/huggingface/huggingface_hub/pull/1111/files).
